docs/client-sdk-states-and-lifecycle.mdx

```diff
@@ -0,0 +1,414 @@
+slug: /client-sdk-states-and-lifecycle
+---
+
+Nimbus SDK Experiment States and Lifecycle
+
+
+###### Authors: Ryan Kelly
+###### Reviewers: Kate Hudson, Tim Smith
+###### Status: Draft as of November 2020; see the [changelog](#changelog) for updates
+
+
+## Introduction
+
+This document provides a high-level overview of the lifecycle of a Nimbus experiment, from the point of view of the client SDK. It's in part an adjunct to the [Nimbus Experiment Publishing Lifecycle](https://github.com/mozilla/experimenter/tree/main/app/experimenter/kinto#experiment-states).
+
+The [behaviour of API calls](#client-behaviors) to the Client SDK will
+depend on both the last observed [server-side experiment
+state](#server-side-experiment-states), and the current [local client
+state](#client-side-experiment-states), and the SDK will [emit
+telemetry](#telemetry) based on the local state of each
+experiment.
+
+I happen to think that these states could map nicely to some Rust Enums, but your mileage may vary.
+
+
+## Key Concepts
+
+
+#### Treatment
+
+We'll use "treatment" to refer to the change in client behaviour that's triggered by an experiment. This might be a small change in colour or wording, or may provide some much larger piece of new UX. The details and scope don't matter for our purposes here, what matters is having a word that means "the user is experiencing the effects of the experiment".
+
+
+#### Enrollment
+
+Enrollment is the process of the client SDK deciding whether this client should be part of the experiment and if so, what branch it should be assigned to. We'll emit an [enrollment](https://mana.mozilla.org/wiki/display/FJT/Nimbus+Engineering#NimbusEngineering-ExperimentTelemetry) telemetry event when this happens.
+
+Once a client is enrolled in an experiment, it will tag its telemetry pings with that experiment and branch information so that they can be identified for analysis.
+
+Enrolled clients will remain enrolled and assigned to the same branch until the experiment ends, or until they reset their experiment state by disabling telemetry. A client may _disable the treatment_ under various circumstances, but it's important that it continue reporting that it was enrolled in the experiment, to [help with data analysis](https://docs.google.com/document/d/18Vdn4zZX2D1u9AqzHRY0lUSUbpBdouqWtVp3Bu8vYQA/).
+
+Note that it's possible for an experiment to be closed for new enrollments, while continuing to be active for and showing its treatment to previously-enrolled clients.
+
+
+#### Exposure
+
+Exposure is the process of actually showing the treatment. In other words, it's when we change client behaviour based on the experiment branch in which the client was enrolled, or when we _would_ have changed behaviour if the user was not in the control group. We'll emit an [exposure](https://mana.mozilla.org/wiki/display/FJT/Nimbus+Engineering#NimbusEngineering-ExperimentTelemetry) telemetry event the first time this happens.
```

Comment:
A comment from @rfk on the original doc:

@k88hudson: IIUC we want to emit an event only the first time the treatment is activated, not every time it's activated.

The latest docs on Mana do not require this to be emitted only once. Is it allowed or perhaps even preferred that we emit this every time the exposure happens?

Reply:
We expose a method that allows it the event to be emitted more than once if desired, depending on the implementation details of the feature (for example, about:newtab might benefit from collecting multiple exposure events for every new tab impression)

Reply:
Yeah, let's treat exposure _events_ as a "once ever" problem and we can implement some labeled counters if we need to get dose information in the future.

Reply:
As far as I can tell, the text here already matches the "once ever" case, so I'm going to mark this as resolved.